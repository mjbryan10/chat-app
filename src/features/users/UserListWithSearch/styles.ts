import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-flow: column nowrap;
  justify-items: center;
  align-items: center;
`;