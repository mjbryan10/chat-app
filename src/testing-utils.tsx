import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }