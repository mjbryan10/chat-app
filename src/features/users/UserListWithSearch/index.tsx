import React, { FC, useState } from 'react';
import { User } from 'shared/Api/types';
import UserList from '../UserList';
import InputTextFieldWithLabel from 'components/InputTextFieldWithLabel';
import * as S from './styles';

interface Props {
   users: User[];
   handleItemClick?: (id: number) => void;
}

const UserListWithSearch: FC<Props> = ({ users, handleItemClick }) => {
   const [searchString, setSearchString] = useState('');
   /**
    * Handler for the search input. Updates the search string state to the
    * user inputted value.
    *
    * @param value String value to set the searchString state with
    */
   const handleSearchChange = (value: string) => {
      setSearchString(value);
   };
   /**
    * @returns An array of users where the username matches the search string
    *
    * @returns Full users array if there is no search string.
    */
   const filteredUsers = () => {
      const usersArray = users;
      if (!searchString.length) return usersArray;
      return usersArray.filter((user) =>
         user.name.toLowerCase().includes(searchString.toLowerCase())
      );
   };
   return (
      <S.Container>
         <InputTextFieldWithLabel
            labelValue="Search"
            disabled={!users.length}
            handleChange={handleSearchChange}
            placeholder="Search for a user"
         />
         <UserList users={filteredUsers()} handleItemClick={handleItemClick} />
      </S.Container>
   );
};

export default UserListWithSearch;
