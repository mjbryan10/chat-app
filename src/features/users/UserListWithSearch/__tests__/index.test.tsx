import React from 'react';
import { render, cleanup, RenderResult, fireEvent } from 'testing-utils';
import UserListWithSearch from '..';

describe('UsersContainer Feature Component', () => {
   const __users = [
      {
         id: 1,
         name: 'Wessel',
      },
      {
         id: 2,
         name: 'Quint',
      },
      {
         id: 3,
         name: 'Mani',
      },
   ];
   const __handleItemClick = jest.fn();
   let utils: RenderResult;
   beforeEach(
      () =>
         (utils = render(
            <UserListWithSearch users={__users} handleItemClick={__handleItemClick} />
         ))
   );
   afterEach(() => {
      cleanup();
   });

   test('should filter the user array to search string by exact match', () => {
      const { getByTestId, getByText, getAllByTestId } = utils;

      fireEvent.change(getByTestId('input-text-field'), { target: { value: 'Wessel' } });

      expect(getByText(/wessel/i)).toBeVisible();
      expect(getAllByTestId('user-item')).toHaveLength(1)
   });

   test('should filter the user array to non-case senstive match', () => {
      const { getByTestId, getByText, getAllByTestId } = utils;

      fireEvent.change(getByTestId('input-text-field'), { target: { value: 'wessel' } });

      expect(getByText(/wessel/i)).toBeVisible();
      expect(getAllByTestId('user-item')).toHaveLength(1)
   });

   test('should filter the user array to partial match', () => {
      const { getByTestId, getByText, getAllByTestId } = utils;

      fireEvent.change(getByTestId('input-text-field'), { target: { value: 'wes' } });

      expect(getByText(/wessel/i)).toBeVisible();
      expect(getAllByTestId('user-item')).toHaveLength(1)
   });

   test('should display the names of users given', () => {
      const { getByText } = utils;

      expect(getByText(/wessel/i)).toBeVisible();
      expect(getByText(/quint/i)).toBeVisible();
      expect(getByText(/mani/i)).toBeVisible();
   });

   test('should call click handler with id of user', () => {
      const { getAllByTestId } = utils;
      const firstUserItem = getAllByTestId('user-item')[0];

      fireEvent.click(firstUserItem);

      expect(__handleItemClick).toBeCalledTimes(1);
      expect(__handleItemClick).toBeCalledWith(__users[0]);
   });
});
