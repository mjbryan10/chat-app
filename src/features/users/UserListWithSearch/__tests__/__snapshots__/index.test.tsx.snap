// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersContainer Feature Component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxhUy cDelPw"
  >
    <label
      class="sc-AxhCb dLEHOo"
    >
      Search
      <input
        class="sc-AxiKw xZNbK"
        data-testid="input-text-field"
        placeholder="Search for a user"
        value=""
      />
    </label>
    <ul
      class="sc-AxirZ kPHIsu"
    >
      <li>
        <div
          class="sc-AxjAm fNKPyA"
          data-testid="user-item"
        >
          Wessel
        </div>
      </li>
      <li>
        <div
          class="sc-AxjAm fNKPyA"
          data-testid="user-item"
        >
          Quint
        </div>
      </li>
      <li>
        <div
          class="sc-AxjAm fNKPyA"
          data-testid="user-item"
        >
          Mani
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
